webbpsf/optics.py

```diff
@@ -2,17 +2,284 @@
+import matplotlib
 
-import astropy.table
+import astropy.table as table
 import astropy.io.fits as fits
+import astropy.units as units
 
 from . import utils
-from .webbpsf_core import _log
 
+import logging
+_log = logging.getLogger('webbpsf')
+
+
+
+#######  Classes for modeling aspects of JWST's segmented active primary #####
+
+def segment_zernike_basis(segnum=1, nterms=15, npix=512, outside = np.nan):
+    """ Basis set in the style of poppy.zernike.zernike_basis for segment-level
+    Zernike polynomials for one segment at a time in JWST's aperture.
+
+    Parameters
+    ------------
+    segnum : integer
+        1 to 18, number of JWST segment. Uses same numbering convention as the WSS.
+    nterms : integer
+        Number of Zernike polynomial terms to return
+    npix : integer
+        Number of pixels per side of the array
+    outside : float
+        Value to fill the array with outside of the valid segment.
+
+    """
+    from .webbpsf_core import segname
+
+    aper = JWST_primary_aperture(label_segments=True)
+    w = poppy.Wavefront(npix=npix,diam=JWST_primary_aperture.circumscribed_diameter)
+    segmask = aper.get_transmission(w)
+
+    segname = segname(segnum)
+    cenx, ceny = aper.seg_centers[segname]
+
+    seg_radius = 1.517/2 # nominal point to point diam for A and B segments;
+                         # ignoring slight departures from ideal hexes for now.
+
+    y,x = w.coordinates()
+
+    r = np.sqrt( (y-ceny)**2 + (x-cenx)**2)/seg_radius
+    theta =  np.arctan2((y-ceny) / seg_radius, (x-cenx) / seg_radius)
+    r[segmask != segnum] = np.nan
+    theta[segmask != segnum] = np.nan
+
+    wg = np.where(segmask==segnum)
+    outzerns = np.full((nterms, npix,npix), outside, dtype=float)
+    outzerns_tmp = poppy.zernike.zernike_basis(nterms=nterms,rho=r[wg], theta=theta[wg], outside=outside)
+    for iz in range(nterms):
+        outzerns[iz][wg] = outzerns_tmp[iz]
+
+    return outzerns
+
+
+class JWST_primary_aperture(poppy.AnalyticOpticalElement):
+    """ The JWST telescope primary mirror geometry, in all its
+    hexagonal obscured complexity. Note this has **just the aperture shape**
+    and not any wavefront error terms.
+
+    JWST design pupil geometry and segment coordinates taken
+    from Paul Lightsey's spreadsheet: "2010.03.16 Transmission X Area Budget.xls".
+    That document was in turn based on Ball Aerospace drawing 2220169 Rev B,
+    and the OTE Cryogenic Optics ICD, BATC doc # C327693.
+
+
+    This class has no wavefront errors, it's just the pupil geometry including
+    the segments (which are not quite perfect hexagons for manufacturing reasons
+    related to trying to tile a curved surface with hexagons while maintaining
+    uniform clearance between adjacent segments) and the secondary mirror support
+    struts, including the bumps on the +V3 strut for the mid boom hinge and mag dampers.
+
+
+    *** CAUTIONARY NOTE ***: At high sampling factors, PSF calculations become a LOT slower.
```

Comment:
I'm not sure if we're following a docstring style, but I favor "NumPy style". I'd change this to

```rest
.. warning:: At high sampling factors, PSF calculations become a LOT slower.
```

(ref: https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#other-points-to-keep-in-mind)